catalog: add implicit record types for all tables

jordanlewis opened the pull request:
Fixes #70036.

This commit adds a resolvable type with the same name as all tables. The
type is treated as a tuple of all of the non-hidden, non-inaccessible
column types of the input table, in creation order.

The table record type is implemented with a virtual type descriptor that
doesn't have an embedded protobuf type descriptor at all.

The new record types are not usable in table definitions for now, but
there's nothing theoretical stopping this from happening. They get
serialized with the `@<int>` format, so it's "just" a matter of teaching
the validators to deal with UDTs sometimes being physical (a real
TypeDescriptor in system.descriptor for the OID - 100000) and sometimes
virtual (a TableDescriptor for that OID).

Release note (sql change): introduce an implicitly-defined type for
every table, which resolves to a tuple type that contains all of the
columns in the table.

jordanlewis committed:
catalog: add implicit record types for all tables

This commit adds a resolvable type with the same name as all tables. The
type is treated as a tuple of all of the non-hidden, non-inaccessible
column types of the input table, in creation order.

The table record type is implemented with a virtual type descriptor that
doesn't have an embedded protobuf type descriptor at all.

The new record types are not usable in table definitions for now, but
there's nothing theoretical stopping this from happening. They get
serialized with the `@<int>` format, so it's "just" a matter of teaching
the validators to deal with UDTs sometimes being physical (a real
TypeDescriptor in system.descriptor for the OID - 100000) and sometimes
virtual (a TableDescriptor for that OID).

Release note (sql change): introduce an implicitly-defined type for
every table, which resolves to a tuple type that contains all of the
columns in the table.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/70100)
<!-- Reviewable:end -->

ajwerner commented on pkg/sql/catalog/typedesc/type_desc.go:
```diff
@@ -984,3 +1012,51 @@ func GetTypeDescriptorClosure(typ *types.T) (map[descpb.ID]struct{}, error) {
+	}
+	return NewBuilder(&descpb.TypeDescriptor{
+		Name:             descriptor.GetName(),
+		ID:               descpb.ID(typeOID),
```
This feels wrong and a likely source of subtle bugs. For enum and alias types, the id is their descriptor ID.

jordanlewis replied:
This is deliberate. If we kept the TypeDescriptor's ID the same as the TableDescriptor's, it would overwrite the TableDescriptor in various caches, and vice versa.

ajwerner replied:
> This is deliberate. If we kept the TypeDescriptor's ID the same as the TableDescriptor's, it would overwrite the TableDescriptor in various caches, and vice versa.

With this approach, that won't happen until descriptor ID `100000` exists, which is something that can very much happen.

jordanlewis replied:
I changed this to be the table ID.

ajwerner reviewed (commented):
This is a neat prototype. I want to make sure we aren't painting ourselves into a corner here. I'd like us to actually make a plan for how we deal with the IDs. For example, for the enums we allocate a distinct oid for the array alias. 

I also have some concerns with hydration and leasing interactions. I have existing concerns about the array alias hydration being wrong. I'd like to see testing to ensure that when we have a tuple type with an enum type and the version on the enum type changes that we properly detect that the query is stale and same goes for the array alias.

jordanlewis commented:
I think the difference between this and for enums is that we *do* actually allocate a real ID for this too. It's the table ID itself.

It's true that the "array aliases" are separately allocated... but the ID of the enum itself is in the descriptor ID space, whereas its OID is in the custom OID space. Here, the ID of the type itself is in the descriptor ID space (it's just the table ID), and it's OID is in the custom OID space as well. I think it's analogous.

I'm realizing that this does not allocate a separate ID for the "array alias" of the table's UDT though. There is no way to solve that without making every new table request 2 new descriptors, which feels wrong. I'd prefer to just fail compatibility with the UDT's array alias. We can see what SQL-E says. Note that we already fail to create that UDT array alias even though we fake compatibility in the pg_type table for the table UDTs themselves.

ajwerner commented:
>making every new table request 2 new descriptors, which feels wrong.

Hard disagree. It should allocate *3* IDs, one for its table, one for its type and one for its type alias. We need to actually design record types and think through these things rather than merging a PR that did something easy at the time and pay for it for the next N years. Let's think through what the design should be and weigh tradeoffs like we usually do for projects.

jordanlewis commented:
As discussed offline, not intending to urgently merge this. We can consider it a prototype for discussion.

I'm not sure if it's necessary to think through persisted record types from end to end to get this in, but maybe we do?

ajwerner commented:
> I'm not sure if it's necessary to think through persisted record types from end to end to get this in, but maybe we do?

I'm inclined to disagree until I can be convinced that what's happening here is both a good idea and something that fits with where we'd want to go. 

I'm also fearful of use of these tuple types in views and all of the column expressions. We should test that heavily.

jordanlewis commented:
> I'm inclined to disagree until I can be convinced that what's happening here is both a good idea and something that fits with where we'd want to go.

Fair enough. I did think through this approach and didn't see anything that would block any future plans. I can write some paragraphs on alternatives if you think it would help.

> I'm also fearful of use of these tuple types in views and all of the column expressions. We should test that heavily.

Agreed. There are some light tests for that in the PR and I'll add some more.

> I'd like to see testing to ensure that when we have a tuple type with an enum type and the version on the enum type changes that we properly detect that the query is stale and same goes for the array alias.

Is there a test that's doing something similar I could crib? Any pointers? Agreed this seems like a ripe area to probe and I'm happy to do so.

ajwerner commented:
Postgres allocates distinct oids for the record types. This is important at least in part because you can grant privileges to the things separately. Also, the usage and reference rules might be nice to separate. 

I see the argument for extreme pragmatism to get compatibility with the tools. In particular, if these things never get serialized anywhere, then maybe it's okay that the implementation is simplistic. Later in the cycle I might be persuaded to take on that debt in the name of progress. Even then, I think we'd want to really carve out code concepts regarding synthetic record types; this here is way implicit in a way that makes me uncomfortable. I'd imagine we'd want a different implementation of the type descriptor interface as opposed to synthesizing an actually bogus type descriptor with a bogus ID. We're also going to have to feel our way around the whole collision at 100000 thing that makes me rather uneasy, especially in a world where creating descriptors becomes as cheap as I think it will be in 22.1. 

As it stands, my preference would be to just decide we want to do record types and do them. Even that project has stages. It does lead me to wonder (yet again) whether the namespace table is structured as we would want it to be. I wonder often about creating a new new namespace table that has a relation type in it so that we could find things like types, functions, and indexes more easily. The last thing I'm worried about it allocating IDs for those things.

postamar commented:
Completely agree with @ajwerner. Let's do these record types properly if we actually need them:
1. We're going to want to have them soon enough anyway, might as well do it right the first time around if this isn't urgent.
2. We're paying down similar tech debt right now with giving the public schema an explicit descriptor and the cost for that is quite high, as it turns out. 
3. Types have enough existing tech debt as it is, see https://github.com/cockroachdb/cockroach/issues/63161 among others. Implementing record types will force some much-needed cleanups so that we don't end up painting ourselves into a corner..

jordanlewis commented:
I pushed an update to this to resolve the outstanding comments from Andrew, add more tests, and remove the issue with collision at 100000 by having the descriptor have the same ID as the table ID.

I'd be fine not merging this, FWIW - I'm learning by doing, and I hope it's not too annoying to talk about different approaches in this way.

I think from a prioritization perspective, if we can make time to do record types end to end somehow in the next release, I'd buy that it'd be better to do that, just as you say. If we can't, maybe there would be a way that this approach could be cleaned up sufficiently to be palatable and not too debt-incurring, by taking some of the suggestions above.

I worry a little bit that it might be very hard to get record types end to end done, because of the idea of doing another new namespace table and another migration. Judging from past experience, that migration could be very heavyweight, difficult and painful... that being said, the lack of it, on its own, is Postgres incompatible of course because we just have the 1 giant namespace and Postgres has 1 per relation kind.

Anyway, just food for thought! Thanks for letting me experiment...

jordanlewis commented:
(btw, since I didn't mention this in the PR description, this is all in pursuit of helping resolve #69010)

jordanlewis commented:
I've redone the PR to have a "virtual descriptor" with no backing protobuf type desc. For now, it's in a separate commit for ease of reviewing; I'll squash once we get closer to deciding what method we want to use here.

```
This commit updates the table record type prototype to use a virtual
type descriptor that doesn't have an embedded protobuf type descriptor
at all.

There are a few TODOs that I'd like review input on.

The main decision is whether or not we should be embedding the
catalog.TableDescriptor that the virtual type descriptor is aliasing.
The commit currently does that, but I could see arguments for why we'd
want to denormalize the fields into the virtual type descriptor struct.

Not sure what we want to do philosophically, so leaving it as an open
question.
```

> I'd like to see testing to ensure that when we have a tuple type with an enum type and the version on the enum type changes that we properly detect that the query is stale and same goes for the array alias.

Testing this, it seems like it doesn't even work for ordinary prepared statements against tables with enum types, so I've filed a separate issue for that: #70378

ajwerner reviewed (commented):
Thanks much for working through this! I'll give it a deeper review another time with fresh eyes. On the whole, this feels close. 

> Reviewed 1 of 6 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-:-MkF_yCm3eRMNxV4IWLX:bglom1v)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jordanlewis)

___
*[pkg/sql/catalog/descs/object.go, line 109 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkFWhSL7OnF2Nna93-3:-MkFWhSL7OnF2Nna93-4:b-zfcx3a) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0ad584c82b986df0763ef717677fe8e70cc820e/pkg/sql/catalog/descs/object.go#L109)):*
> ```Go
> 			// created record type for the table that we found.
> 			if flags.RequireMutable {
> 				desc = typedesc.CreateMutableFromTableDesc(tableDesc)
> ```

Should this return an error if the client requested that the type be mutable here? We do it one level up on the `ByID` path but I think you have the right info here for this path. 
___
*[pkg/sql/catalog/typedesc/type_desc.go, line 1027 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkF_fzm-6aJGx7hjh1v:-MkF_fzm-6aJGx7hjh1w:b-ivi4dy) ([raw file](https://github.com/cockroachdb/cockroach/blob/54cb2ad83b973fc1f04efef83cba6c4ffce9d223/pkg/sql/catalog/typedesc/type_desc.go#L1027)):*
> ```Go
> // casted to a table type. This is important for Postgres compatibility, which
> // demands that every table have such a corresponding type.
> func descBuilderFromTableDesc(descriptor catalog.TableDescriptor) TypeDescriptorBuilder {
> ```

should we assert that this table descriptor is indeed already hydrated somehow? We really want it to be by the time we're here, no?


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70100)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jordanlewis)

___
*[pkg/sql/catalog/typedesc/type_desc.go, line 1027 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkF_fzm-6aJGx7hjh1v:-MkZtNqB-XwGpDTO4zOH:b8ufav9) ([raw file](https://github.com/cockroachdb/cockroach/blob/54cb2ad83b973fc1f04efef83cba6c4ffce9d223/pkg/sql/catalog/typedesc/type_desc.go#L1027)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

should we assert that this table descriptor is indeed already hydrated somehow? We really want it to be by the time we're here, no?
</blockquote></details>

This method isn't used any longer and I deleted it, but your comment still applies to the new method below. That being said I'm not really sure 
___
*[pkg/sql/catalog/typedesc/type_desc.go, line 1077 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkZthRr9NgoXACWHAYh:-MkZthRr9NgoXACWHAYi:b-gmm9mj) ([raw file](https://github.com/cockroachdb/cockroach/blob/54cb2ad83b973fc1f04efef83cba6c4ffce9d223/pkg/sql/catalog/typedesc/type_desc.go#L1077)):*
> ```Go
> type VirtualTableRecordType struct {
> 	// desc is the TableDescriptor that this virtual record type is created from.
> 	// TODO(jordan): does it make more sense to eagerly pop out all the fields
> ```

Thoughts on this from reviewers?
___
*[pkg/sql/catalog/typedesc/type_desc.go, line 1085 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkZtWkK0xGS6GWyXJ2Y:-MkZtWkK0xGS6GWyXJ2Z:byxj6eu) ([raw file](https://github.com/cockroachdb/cockroach/blob/54cb2ad83b973fc1f04efef83cba6c4ffce9d223/pkg/sql/catalog/typedesc/type_desc.go#L1085)):*
> ```Go
> }
> 
> // TODO(jordan): if we keep the embedded catalog.TableDescriptor, do we want
> ```

Thoughts on this from reviewers? There's a ton of "pass through" delegation in this file, so it's certainly fewer LOC to embed, but slightly nervous about bugs from not considering cases if we do that?


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70100)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jordanlewis)

___
*[pkg/sql/catalog/typedesc/type_desc.go, line 1027 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkF_fzm-6aJGx7hjh1v:-MkZuqgNDTEqCmfaBbOI:b-tnr6kt) ([raw file](https://github.com/cockroachdb/cockroach/blob/54cb2ad83b973fc1f04efef83cba6c4ffce9d223/pkg/sql/catalog/typedesc/type_desc.go#L1027)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This method isn't used any longer and I deleted it, but your comment still applies to the new method below. That being said I'm not really sure 
</blockquote></details>

Whoops, unfinished comment - I added the validation.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
Seems like this needs some logic test changes to deal with error message changes. This feels pretty close. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-:-MkbZeOrDnJSiVxqqcb9:b-hpikp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jordanlewis)

___
*[pkg/sql/catalog/typedesc/type_desc.go, line 1027 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkbZRTy7kShRst07V_Y:-MkbZRTy7kShRst07V_Z:b5x20m6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9fac8141ae73e3fac738a2cc222a7a0d9f29dc5/pkg/sql/catalog/typedesc/type_desc.go#L1027)):*
> ```Go
> 	// rather than have all of the methods delegate?
> 	desc catalog.TableDescriptor
> ```

I have a slight preference to pop the fields out in the constructor. What do you think? 
___
*[pkg/sql/catalog/typedesc/type_desc.go, line 1029 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-Mk_LEZ-ChADNWtTp4Y-:-Mk_LEZ-ChADNWtTp4Y0:b-zd2bwn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9fac8141ae73e3fac738a2cc222a7a0d9f29dc5/pkg/sql/catalog/typedesc/type_desc.go#L1029)):*
<details><summary><i alt="summary">Quoted 12 lines of code…</i></summary>

> ```
> // VirtualTableRecordType is an implementation of catalog.TypeDescriptor that
> // represents a record type for a particular table: meaning, the composite type
> // that contains, in order, all of the visible columns for the table.
> type VirtualTableRecordType struct {
> 	// desc is the TableDescriptor that this virtual record type is created from.
> 	// TODO(jordan): does it make more sense to eagerly pop out all the fields
> 	// that we care about from this backing TableDescriptor into the struct,
> 	// rather than have all of the methods delegate?
> 	desc catalog.TableDescriptor
> 
> 	typ *types.T
> }
> ```
</details>

nit: pull this struct, its constructor and all of its logic into a separate file 
___
*[pkg/sql/catalog/typedesc/type_desc.go, line 1034 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkbZOlvAzlIgXSXe0MM:-MkbZOlvAzlIgXSXe0MN:b-339gox) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9fac8141ae73e3fac738a2cc222a7a0d9f29dc5/pkg/sql/catalog/typedesc/type_desc.go#L1034)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> // TODO(jordan): if we keep the embedded catalog.TableDescriptor, do we want
> // to just use Go embedding to avoid writing all of these delegate calls? For now,
> // I like keeping them explicit, since it forces us to think about the implementations
> // separately from the table descriptor itself.
> ```
</details>

I like this. 
___
*[pkg/sql/catalog/typedesc/type_desc.go, line 1071 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkbYe4N2DuOg-d5U3l4:-MkbYe4N2DuOg-d5U3l5:b13789n) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9fac8141ae73e3fac738a2cc222a7a0d9f29dc5/pkg/sql/catalog/typedesc/type_desc.go#L1071)):*
> ```Go
> 	// TODO(jordan): we can't drop these virtual types, so is it weird to say
> 	// that we have drop permission on this thing even if we don't?
> 	return v.desc.GetPrivileges()
> ```

Should we mask this to just the type privileges (which I think is just usage?) Should we instead map `select` to `usage`? 
___
*[pkg/sql/catalog/typedesc/type_desc.go, line 1088 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-Mk_L3O5C14MIVD2uW1C:-Mk_L3O65_WdHyTF7xB_:b7ytotw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9fac8141ae73e3fac738a2cc222a7a0d9f29dc5/pkg/sql/catalog/typedesc/type_desc.go#L1088)):*
> ```Go
> 
> // Adding implements the Descriptor interface.
> func (v VirtualTableRecordType) Adding() bool { panic("can't call adding on VirtualTableRecordType") }
> ```

thoughts on adding a method that these string panics delegate to which add info about the descriptor itself and wraps it in an assertion failed error before panicking? 
___
*[pkg/sql/catalog/typedesc/type_desc.go, line 1246 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-Mk_LO3a6RFTm9PHjcHM:-Mk_LO3a6RFTm9PHjcHN:baohgg3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9fac8141ae73e3fac738a2cc222a7a0d9f29dc5/pkg/sql/catalog/typedesc/type_desc.go#L1246)):*
> ```Go
> // the implicit record type for a table, which has 1 field for every visible
> // column in the table.
> func CreateVirtualRecordTypeFromTableDesc(
> ```

nit: pull the constructor up above the methods. 


<!-- Sent from Reviewable.io -->

jordanlewis committed:
logictest: move json populate record tests

Mechanical change to move the json populate record tests to the
json_builtins file from the json file, for organization purposes.

Release note: None

jordanlewis committed:
builtins: test anon table type with json_populate

And fix a small missing error message.

Release note: None

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70100)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jordanlewis)

___
*[pkg/sql/catalog/typedesc/type_desc.go, line 1034 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkbZOlvAzlIgXSXe0MM:-Mkb_HLiEHpVfFTi3zaC:bvlxtps) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9fac8141ae73e3fac738a2cc222a7a0d9f29dc5/pkg/sql/catalog/typedesc/type_desc.go#L1034)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> // TODO(jordan): if we keep the embedded catalog.TableDescriptor, do we want
> // to just use Go embedding to avoid writing all of these delegate calls? For now,
> // I like keeping them explicit, since it forces us to think about the implementations
> // separately from the table descriptor itself.
> ```
I like this. 
</blockquote></details>

Not sure which alternative you are referring to with "this" - I'm guessing you're saying you like the way it is?
___
*[pkg/sql/catalog/typedesc/type_desc.go, line 1071 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkbYe4N2DuOg-d5U3l4:-MkeWi2E-Gz1TlSuAz1Y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9fac8141ae73e3fac738a2cc222a7a0d9f29dc5/pkg/sql/catalog/typedesc/type_desc.go#L1071)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Should we mask this to just the type privileges (which I think is just usage?) Should we instead map `select` to `usage`? 
</blockquote></details>

Done.
___
*[pkg/sql/catalog/typedesc/type_desc.go, line 1088 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-Mk_L3O5C14MIVD2uW1C:-MkgoUAi4wo-yL-oUcqf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9fac8141ae73e3fac738a2cc222a7a0d9f29dc5/pkg/sql/catalog/typedesc/type_desc.go#L1088)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

thoughts on adding a method that these string panics delegate to which add info about the descriptor itself and wraps it in an assertion failed error before panicking? 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70100)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/sql/catalog/typedesc/type_desc.go, line 1027 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkbZRTy7kShRst07V_Y:-MkgomO17poJIj8UB3YF:bu2d69u) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9fac8141ae73e3fac738a2cc222a7a0d9f29dc5/pkg/sql/catalog/typedesc/type_desc.go#L1027)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I have a slight preference to pop the fields out in the constructor. What do you think? 
</blockquote></details>

I guess it feels duplicative to pop the fields out in a way that slightly scares me - is there a chance of code migrations not hitting the set of fields we pop out, or something? I dunno.


<!-- Sent from Reviewable.io -->

jordanlewis commented:
Sending a heads up to @rafiss to please take a look at this, since I know you've been looking at tuple type related things.

rafiss commented:
just saw that this landed in PostgreSQL 14. probably not relevant for what we want to do right now, but sharing for awareness
```
Create composite array types for system catalogs (Wenjing Zeng)

User-defined relations have long had composite types associated with them, and also array types over those composite types. System catalogs now do as well. This change also fixes an inconsistency that creating a user-defined table in single-user mode would fail to create a composite array type.
```

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70100)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jordanlewis, and @rafiss)

___
*[pkg/sql/catalog/descpb/structured.proto, line 1371 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkxWAQbEv6huzTn7h6e:-MkxWAQbEv6huzTn7h6f:b8o7i9j) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f26602c9ecede53de9245fd33171227a35998ee/pkg/sql/catalog/descpb/structured.proto#L1371)):*
> ```protobuf
>     // kind of TypeDescriptor is *never* persisted to disk! If you are here,
>     // thinking about using or persisting this value, you should *not* do that!
>     VIRTUAL_TABLE_RECORD_TYPE = 3;
> ```

this reads like it is the "record type for virtual tables"

would it be more accurate to name this `TABLE_VIRTUAL_RECORD_TYPE` or `TABLE_RECORD_VIRTUAL_TYPE`?
___
*[pkg/sql/catalog/typedesc/virtual_table_record_type.go, line 36 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkxXR0M8c3qjysJAghb:-MkxXR0M8c3qjysJAghc:bdgmim2) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f26602c9ecede53de9245fd33171227a35998ee/pkg/sql/catalog/typedesc/virtual_table_record_type.go#L36)):*
> ```Go
> 
> 	typ   *types.T
> 	privs *descpb.PrivilegeDescriptor
> ```

maybe add a comment saying what privileges are added?
___
*[pkg/sql/catalog/typedesc/virtual_table_record_type.go, line 38 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkxXKuK0deDmM9021JK:-MkxXKuK0deDmM9021JL:bad0z83) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f26602c9ecede53de9245fd33171227a35998ee/pkg/sql/catalog/typedesc/virtual_table_record_type.go#L38)):*
> ```Go
> 	privs *descpb.PrivilegeDescriptor
> }
> ```

what would we need to handle the array type for this virtual record type?
___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 3408 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkxXp_Z1xRfUUW2TkPV:-MkxXp_Z1xRfUUW2TkPW:bi8i9kr) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f26602c9ecede53de9245fd33171227a35998ee/pkg/sql/logictest/testdata/logic_test/pg_catalog#L3408)):*
> ```
> SELECT oid
> FROM pg_catalog.pg_type
> WHERE typname = 'source_table'
> ```

why use `let` here instead of combining it into one

``` select ... FROM pg_type where typname = 'source_table'```
___
*[pkg/sql/sem/tree/casts.go, line 375 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkxYh5J1_mGSb_ktSji:-MkxYh5J1_mGSb_ktSjj:b-puk5i6) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f26602c9ecede53de9245fd33171227a35998ee/pkg/sql/sem/tree/casts.go#L375)):*
> ```Go
> 	// Casts to TupleFamily.
> 	{from: types.UnknownFamily, to: types.TupleFamily, volatility: VolatilityImmutable},
> 	{from: types.TupleFamily, to: types.TupleFamily, volatility: VolatilityImmutable},
> ```

hm, is it really immutable? does it show up in the postgresql pg_cast table?

what about casting a tuple containing a timestamptz to a tuple containing a string? that can't be immutable right?


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70100)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jordanlewis, and @rafiss)

___
*[pkg/sql/sem/tree/casts.go, line 1449 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkxZXIHEmmqdj0gthkt:-MkxZXIHEmmqdj0gthku:b-eqy2jk) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f26602c9ecede53de9245fd33171227a35998ee/pkg/sql/sem/tree/casts.go#L1449)):*
> ```Go
> 			return ParseDOid(ctx, string(*v), t)
> 		}
> 	case types.TupleFamily:
> ```

i feel like we'd want handling for casting to arrays of tuples too. or at least tests for it, to make sure the error message is OK


<!-- Sent from Reviewable.io -->

rafiss commented:
cc @otan as well (in case you're interested) since the story of getting the right IDs for this stuff was previously explored in https://github.com/cockroachdb/cockroach/pull/66815

postamar commented on pkg/sql/catalog/descpb/structured.proto:
```diff
@@ -1365,6 +1365,10 @@ message TypeDescriptor {
+    // Represents the virtual record type created on behalf of each table. This
+    // kind of TypeDescriptor is *never* persisted to disk! If you are here,
+    // thinking about using or persisting this value, you should *not* do that!
+    VIRTUAL_TABLE_RECORD_TYPE = 3;
```
I don't hate it, but "virtual" may be a bit overloaded. What about "transient" instead? or "implicit"?

postamar commented on pkg/sql/catalog/typedesc/type_desc.go:
```diff
@@ -513,6 +513,8 @@ func (desc *immutable) ValidateSelf(vea catalog.ValidationErrorAccumulator) {
 			vea.Report(errors.AssertionFailedf("ALIAS type desc has array type ID %d", desc.GetArrayTypeID()))
 		}
+	case descpb.TypeDescriptor_VIRTUAL_TABLE_RECORD_TYPE:
+		vea.Report(errors.AssertionFailedf("cannot persist type descriptor kind %s", desc.Kind.String()))
```
I find this message confusing. For a start we shouldn't assume that validation only happens during descriptor reads and writes. Really if we believe that a type descriptor of this kind should never even be validated in the first place then that's what the message should read, accompanied with an explanation as to its transient nature.

postamar replied:
I would also prefer for this validation failure to be exercised in a test.

postamar commented on pkg/sql/catalog/typedesc/virtual_table_record_type.go:
```diff
@@ -0,0 +1,316 @@
+	// desc is the TableDescriptor that this virtual record type is created from.
+	// TODO(jordan): does it make more sense to eagerly pop out all the fields
+	// that we care about from this backing TableDescriptor into the struct,
+	// rather than have all of the methods delegate?
```
My personal answer to that question would be: no, keep on delegating. From the point of view of the lifecycle of a record type descriptor, it's clear that its underlying the table descriptor will never change for the duration of its existence. Plus they're so tightly coupled.

postamar commented on pkg/sql/catalog/typedesc/virtual_table_record_type.go:
```diff
@@ -0,0 +1,316 @@
+// VirtualTableRecordType is an implementation of catalog.TypeDescriptor that
+// represents a record type for a particular table: meaning, the composite type
+// that contains, in order, all of the visible columns for the table.
+type VirtualTableRecordType struct {
```
Is there a reason why this isn't package-private?

postamar commented on pkg/sql/catalog/typedesc/virtual_table_record_type.go:
```diff
@@ -0,0 +1,316 @@
+func (v VirtualTableRecordType) GetDrainingNames() []descpb.NameInfo {
+	// TODO(jordan): I don't think we need to return anything here, because this
+	// descriptor doesn't really "own" any names on top of what the table already
+	// does.
```
This is correct plus we're getting rid of draining names.

postamar commented on pkg/sql/catalog/typedesc/virtual_table_record_type.go:
```diff
@@ -0,0 +1,316 @@
+
+// GetIDClosure implements the TypeDescriptor interface.
+func (v VirtualTableRecordType) GetIDClosure() (map[descpb.ID]struct{}, error) {
+	ret := make(map[descpb.ID]struct{})
```
Should `v.GetID()` be included in the ID closure? I'm guessing perhaps not, but I'm asking to make sure that this has been considered.

postamar commented on pkg/sql/catalog/typedesc/virtual_table_record_type.go:
```diff
@@ -0,0 +1,316 @@
+
+// GetReferencedDescIDs implements the Descriptor interface.
+func (v VirtualTableRecordType) GetReferencedDescIDs() (catalog.DescriptorIDSet, error) {
+	return v.desc.GetReferencedDescIDs()
```
I don't think this matters much yet, but this isn't accurate. Presently this will include sequence references, foreign key references, etc. Instead this should be `GetIDClosure()` supplemented with `v.GetID()`, `v.GetParentID()` and `v.GetParentSchemaID()`.

postamar commented on pkg/sql/catalog/validate.go:
```diff
@@ -395,7 +396,15 @@ func (vdg *validationDescGetterImpl) GetTypeDescriptor(id descpb.ID) (TypeDescri
-	return AsTypeDescriptor(desc)
+	if _, ok := desc.(TableDescriptor); ok {
+		return nil, unimplemented.NewWithIssuef(70099, "table record type %q not allowed in definition",
+			desc.GetName())
```
This makes me uneasy. We should not be assuming that the descriptor holding the reference intended that ID to be to a record type descriptor. It could be the result of just about any kind of descriptor corruption, we don't know.

Unless this serves a purpose I'm not clued in to, I'd prefer to remove this if-block entirely.

postamar reviewed (changes requested):
I found a few little things to complain about but I'm generally happy with how this is turning out.

jordanlewis reviewed (commented):
TFTRs, this is RFAL!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-:-MlWyHr48WgdG78Hgu26:b-xuvn8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @postamar, and @rafiss)

___
*[pkg/sql/catalog/validate.go, line 401 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MjSD9HfA2o67HnbbwjI-r15-401:-MlWvz9NE7BejjWl3or2:b7n7m08) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed9e8672f8b68f4a83624971273612e576ae07ac/pkg/sql/catalog/validate.go#L401)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This makes me uneasy. We should not be assuming that the descriptor holding the reference intended that ID to be to a record type descriptor. It could be the result of just about any kind of descriptor corruption, we don't know.

Unless this serves a purpose I'm not clued in to, I'd prefer to remove this if-block entirely.
</blockquote></details>

This is a good point. It's defensive programming. If we had an issue, and we did have some kind of weird corruption, we'd get this unimplemented error instead of a crash. But, we'd probably prefer a crash here, I agree. Done.
___
*[pkg/sql/catalog/descpb/structured.proto, line 1371 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkxWAQbEv6huzTn7h6e:-MlWgF4_6z5SLSEsod9Y:bpkmcgi) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f26602c9ecede53de9245fd33171227a35998ee/pkg/sql/catalog/descpb/structured.proto#L1371)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this reads like it is the "record type for virtual tables"

would it be more accurate to name this `TABLE_VIRTUAL_RECORD_TYPE` or `TABLE_RECORD_VIRTUAL_TYPE`?
</blockquote></details>

I've renamed to TABLE_IMPLICIT_RECORD_TYPE.
___
*[pkg/sql/catalog/descpb/structured.proto, line 1371 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-Mk_KFxjBxknzmyl7Pdi-r15-1371:-MlWgHh9BHAa_XcGsayc:bpkmcgi) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed9e8672f8b68f4a83624971273612e576ae07ac/pkg/sql/catalog/descpb/structured.proto#L1371)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I don't hate it, but "virtual" may be a bit overloaded. What about "transient" instead? or "implicit"?
</blockquote></details>

I've renamed to TABLE_IMPLICIT_RECORD_TYPE.
___
*[pkg/sql/catalog/typedesc/type_desc.go, line 517 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MjSD9HfA2o67HnbbwjH-r15-517:-MlWlq7NBh65Eioi_mcp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed9e8672f8b68f4a83624971273612e576ae07ac/pkg/sql/catalog/typedesc/type_desc.go#L517)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I would also prefer for this validation failure to be exercised in a test.
</blockquote></details>

Done.
___
*[pkg/sql/catalog/typedesc/virtual_table_record_type.go, line 36 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkxXR0M8c3qjysJAghb:-MlWmbFU8_Ur3KFYeXYh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f26602c9ecede53de9245fd33171227a35998ee/pkg/sql/catalog/typedesc/virtual_table_record_type.go#L36)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

maybe add a comment saying what privileges are added?
</blockquote></details>

Done.
___
*[pkg/sql/catalog/typedesc/virtual_table_record_type.go, line 38 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkxXKuK0deDmM9021JK:-MlWnFbnDKMy3JyIHkQr:b-z6c7pn) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f26602c9ecede53de9245fd33171227a35998ee/pkg/sql/catalog/typedesc/virtual_table_record_type.go#L38)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

what would we need to handle the array type for this virtual record type?
</blockquote></details>

The reason we're not doing this here is that we'd have to allocate a new descriptor ID to keep "safe" from any new schema objects being created.

Let's say we were wanting to keep the array type oids next to the type oids. Then, for a table id 52, for which we have a corresponding type id 100052, we'd need a corresponding array type id 100053, or something, but we have no way to reserve that ID right now. It's a relatively straightforward extension, but I figured we could get to it later since this PR doesn't close any doors that prevent us from getting there.
___
*[pkg/sql/catalog/typedesc/virtual_table_record_type.go, line 28 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkecIJhqIIrD-kKMvyp-r15-28:-MlWlwdC24Cd0HIe9rh4:b-v0p89) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed9e8672f8b68f4a83624971273612e576ae07ac/pkg/sql/catalog/typedesc/virtual_table_record_type.go#L28)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Is there a reason why this isn't package-private?
</blockquote></details>

Yes, we have a couple of type-checks for it in the resolution path.
___
*[pkg/sql/catalog/typedesc/virtual_table_record_type.go, line 32 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkecIJhqIIrD-kKMvyp-r15-32:-MlWm5MeEwMBGoEP9KwP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed9e8672f8b68f4a83624971273612e576ae07ac/pkg/sql/catalog/typedesc/virtual_table_record_type.go#L32)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

My personal answer to that question would be: no, keep on delegating. From the point of view of the lifecycle of a record type descriptor, it's clear that its underlying the table descriptor will never change for the duration of its existence. Plus they're so tightly coupled.
</blockquote></details>

Done.
___
*[pkg/sql/catalog/typedesc/virtual_table_record_type.go, line 81 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkecIJhqIIrD-kKMvyp-r15-81:-MlWnjHFDRHtRwZxUXq6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed9e8672f8b68f4a83624971273612e576ae07ac/pkg/sql/catalog/typedesc/virtual_table_record_type.go#L81)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

`if privilege.SELECT.IsSetIn(tablePrivs.Users[i].Privileges) {`
</blockquote></details>

Done.
___
*[pkg/sql/catalog/typedesc/virtual_table_record_type.go, line 82 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkecIJhqIIrD-kKMvyp-r15-82:-MlWnm_L5XMHJq43dUy0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed9e8672f8b68f4a83624971273612e576ae07ac/pkg/sql/catalog/typedesc/virtual_table_record_type.go#L82)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

`newPrivs[i].Privileges = privilege.USAGE.Mask()`
</blockquote></details>

Done.
___
*[pkg/sql/catalog/typedesc/virtual_table_record_type.go, line 124 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkecIJhqIIrD-kKMvyp-r15-124:-MlWnrdx1akNaFQ0wsc3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed9e8672f8b68f4a83624971273612e576ae07ac/pkg/sql/catalog/typedesc/virtual_table_record_type.go#L124)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This is correct plus we're getting rid of draining names.
</blockquote></details>

Done.
___
*[pkg/sql/catalog/typedesc/virtual_table_record_type.go, line 178 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkecIJhqIIrD-kKMvyp-r15-178:-MlWo-4j-M0FT7C0EgY8:bnchgf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed9e8672f8b68f4a83624971273612e576ae07ac/pkg/sql/catalog/typedesc/virtual_table_record_type.go#L178)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I don't think this matters much yet, but this isn't accurate. Presently this will include sequence references, foreign key references, etc. Instead this should be `GetIDClosure()` supplemented with `v.GetID()`, `v.GetParentID()` and `v.GetParentSchemaID()`.
</blockquote></details>

Done. I agree it doesn't matter but I'll keep it correct.

On second thought I've made this return an error; it shouldn't be called from anywhere.jkj
___
*[pkg/sql/catalog/typedesc/virtual_table_record_type.go, line 238 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkecIJhqIIrD-kKMvyp-r15-238:-MlWorJi9xGRU9QujeC7:bmfe8nb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed9e8672f8b68f4a83624971273612e576ae07ac/pkg/sql/catalog/typedesc/virtual_table_record_type.go#L238)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Should `v.GetID()` be included in the ID closure? I'm guessing perhaps not, but I'm asking to make sure that this has been considered.
</blockquote></details>

Yes, it should, you're right, done. I think ultimately this doesn't matter much because we cannot store implicit record table types in any actual serialized tables or types, but again, good to get it correct.

On second thought I've made this return an error; it shouldn't be called from anywhere.
___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 3408 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkxXp_Z1xRfUUW2TkPV:-MlWpsCz3tS-0LuA5awi:btd58ao) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f26602c9ecede53de9245fd33171227a35998ee/pkg/sql/logictest/testdata/logic_test/pg_catalog#L3408)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

why use `let` here instead of combining it into one

``` select ... FROM pg_type where typname = 'source_table'```
</blockquote></details>

The reason for this was that I wanted to exercise the pg_type virtual index as well. Added a comment to explain why it's done this way, as well as another test that's similar just for the non-virtual index case.
___
*[pkg/sql/sem/tree/casts.go, line 375 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkxYh5J1_mGSb_ktSji:-MlWqduzBPx2kiH4L7Nv:b-hdfljn) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f26602c9ecede53de9245fd33171227a35998ee/pkg/sql/sem/tree/casts.go#L375)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hm, is it really immutable? does it show up in the postgresql pg_cast table?

what about casting a tuple containing a timestamptz to a tuple containing a string? that can't be immutable right?
</blockquote></details>

There's actually no cases for 2249 (record) in pg_cast... I'm not sure what that would mean? Clearly we can cast tuple to tuple...

Nevertheless you're right, I've made it VolatilityStable which I think will fix these issues.
___
*[pkg/sql/sem/tree/casts.go, line 1449 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-MkxZXIHEmmqdj0gthkt:-MlWsPDlBtD9iivxljRF:b-duarfy) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f26602c9ecede53de9245fd33171227a35998ee/pkg/sql/sem/tree/casts.go#L1449)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i feel like we'd want handling for casting to arrays of tuples too. or at least tests for it, to make sure the error message is OK
</blockquote></details>

Great point! This was actually broken :flushed: 

I've fixed the issue and added some new tests.


<!-- Sent from Reviewable.io -->

postamar reviewed (approved):
Thanks, I have no further complaints.

ajwerner reviewed (commented):
Nice! Just some minor questions and nits from me.

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70100#-:-MlzbJqG6BIWxzJw30VR:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @jordanlewis, @postamar, and @rafiss)



<!-- Sent from Reviewable.io -->
